### PR TITLE
#1 Add the Executor interface responsible for updating the DB

### PR DESCRIPTION
**Pull request by jon-betts**

### CODE REVIEW

File: h/h_api/bulk_api/executor.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+class Executor:
+    """A callback to provide concrete functionality for actions in BulkJob."""
+
+    def configure(self, config):
+        """
+        Configure the job from the provided config object.
+
+        :param config: A config object
+        :return:
```
Comment (seanh) [inline h/h_api/bulk_api/executor.py]: ?

File: h/h_api/bulk_api/executor.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        """
+        pass
+
+    def execute_batch(self, command_type, data_type, default_config, batch):
+        """
+        Execute the actions specified, returning any referenced ids.
+
+        The implementer is expected to:
+
+         * Full-fill the action
```
Comment (seanh) [inline h/h_api/bulk_api/executor.py]:
```suggestion
         * Execute the actions
```

File: h/h_api/bulk_api/executor.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+         * Respect the configuration passed
+         * Return a list of `Report` objects in the same order as the items
+           in the batch
+
+        :param command_type: The CommandType being (e.g. UPSERT)
```
Comment (seanh) [inline h/h_api/bulk_api/executor.py]:
```suggestion
        :param command_type: The CommandType being executed (e.g. CommandType.UPSERT)
```

File: h/h_api/bulk_api/executor.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        :param data_type: The DataType being modified
```
Comment (seanh) [inline h/h_api/bulk_api/executor.py]:
```suggestion
        :param data_type: The DataType being modified, e.g. DataType.USER
```

File: h/h_api/bulk_api/executor.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        :param default_config: Configuration which applies to all actions
+        :param batch: A list of command objects
```
Comment (seanh) [inline h/h_api/bulk_api/executor.py]:
```suggestion
        :param batch: A list of Command models
```

I think there might be something going on here where the `command.type` of each command in `batch` must be equal to the given `command_type` argument? Also for `data_type`. Might be worth explaining the relationship between these if any

File: h/h_api/bulk_api/executor.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        The implementer is expected to:
+
+         * Full-fill the action
+         * Respect the configuration passed
+         * Return a list of `Report` objects in the same order as the items
+           in the batch
+
+        :param command_type: The CommandType being (e.g. UPSERT)
+        :param data_type: The DataType being modified
+        :param default_config: Configuration which applies to all actions
```
Comment (seanh) [inline h/h_api/bulk_api/executor.py]:
```suggestion
        :param default_config: The Configuration model that applies to all actions
```

File: h/h_api/bulk_api/executor.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        :param batch: A list of command objects
+        :return: A list of `Report` objects
+        """
+        raise NotImplementedError()
+
+    def get_items(self, data_type, ids, config):
+        """
+        Retrieve a number of items to report back to the caller.
```
Comment (seanh) [inline h/h_api/bulk_api/executor.py]:
```suggestion
        Return the items identified by `ids`.
```

I think whether or not these items get reported back to the ultimate caller, is the business of whatever code is calling `get_items()`. `get_items()` itself just returns the items it's asked for

File: h/h_api/bulk_api/executor.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+        The objects returned should be in the same order as sent.
```
Comment (seanh) [inline h/h_api/bulk_api/executor.py]:
```suggestion
        The items returned should be in the same order as requested in `ids`.
```

File: h/h_api/bulk_api/model/report.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    # Attach this here for the convenience of the consuming library
+    CommandResult = CommandResult
+
+    def __init__(self, outcome, id_):
+        """
+        :param outcome: An instance of CommandResult
+        :param id_: The id of the resource update
+        """
+        if id_ is None:
+            raise ValueError("Id is required for successful outcomes")
```
Comment (seanh) [inline h/h_api/bulk_api/model/report.py]:
```suggestion
            raise ValueError("id_ is required for successful outcomes")
```

File: h/h_api/bulk_api/executor.py (hunk 6 above)
Comment (jon-betts) [reply h/h_api/bulk_api/executor.py]: This isn't quite right, it's a dict at this point. I'll change this to reflect that

File: h/h_api/bulk_api/executor.py (hunk 7 above)
Comment (jon-betts) [reply h/h_api/bulk_api/executor.py]: Yup. I agree

File: h/h_api/bulk_api/executor.py (hunk 1 above)
Comment (jon-betts) [reply h/h_api/bulk_api/executor.py]: Looks like a bit of pycharm cruft mixed with some good ole' fashioned Betts vagueness

Comment (jon-betts) [reply h/h_api/bulk_api/executor.py]: I've made this clearer now. This isn't the entire config object, it's only options deemed relevant to this action for the executor (plenty is handled for you).

File: h/h_api/bulk_api/executor.py (hunk 6 above)
Comment (jon-betts) [reply h/h_api/bulk_api/executor.py]: I've made this clearer now. This isn't the entire config object, it's only options deemed relevant to this action for the executor (plenty is handled for you).

**Pull request merged**